tinytuya/BulbDevice.py

```diff
@@ -588,33 +599,33 @@ def set_colourtemp(self, colourtemp, nowait=False):
                 "set_colourtemp: The colour temperature needs to be between 0 and 1000.",
             )
 
-        payload = self.generate_payload(
-            CONTROL, {self.DPS_INDEX_COLOURTEMP[self.bulb_type]: colourtemp}
-        )
-        data = self._send_receive(payload, getresponse=(not nowait))
-        return data
+        return self.set_value( self.DPS_INDEX_COLOURTEMP[self.bulb_type], colourtemp, nowait=nowait )
 
     def brightness(self):
         """Return brightness value"""
-        return self.status()[self.DPS][self.DPS_INDEX_BRIGHTNESS[self.bulb_type]]
+        if not self.bulb_type: self.detect_bulb()
+        return self.cached_status()[self.DPS][self.DPS_INDEX_BRIGHTNESS[self.bulb_type]]
 
     def colourtemp(self):
         """Return colour temperature"""
-        return self.status()[self.DPS][self.DPS_INDEX_COLOURTEMP[self.bulb_type]]
+        if not self.bulb_type: self.detect_bulb()
+        return self.cached_status()[self.DPS][self.DPS_INDEX_COLOURTEMP[self.bulb_type]]
 
     def colour_rgb(self):
         """Return colour as RGB value"""
-        hexvalue = self.status()[self.DPS][self.DPS_INDEX_COLOUR[self.bulb_type]]
+        if not self.bulb_type: self.detect_bulb()
+        hexvalue = self.cached_status()[self.DPS][self.DPS_INDEX_COLOUR[self.bulb_type]]
         return BulbDevice._hexvalue_to_rgb(hexvalue, self.bulb_type)
 
     def colour_hsv(self):
         """Return colour as HSV value"""
-        hexvalue = self.status()[self.DPS][self.DPS_INDEX_COLOUR[self.bulb_type]]
+        if not self.bulb_type: self.detect_bulb()
+        hexvalue = self.cached_status()[self.DPS][self.DPS_INDEX_COLOUR[self.bulb_type]]
         return BulbDevice._hexvalue_to_hsv(hexvalue, self.bulb_type)
 
     def state(self):
         """Return state of Bulb"""
-        status = self.status()
+        status = self.cached_status()

```

Comment:
I like the idea of caching. But what if the user is making changes to the device using the SmartLife app or other tool?

* We should definitely cache the device type and capabilities.
* Status or current state of device should be live (unless we provide a force=False)?

Thoughts?

Reply:
> But what if the user is making changes to the device using the SmartLife app or other tool?

Every device I've seen sends asynchronous updates when that happens, so as long as the user calls `.receive()` every now and then or sends a command with `nowait=False` then those changes will be picked up and cached.  This is also why caching is only enabled when `persist=True` and an active connection to the device is open.

Reply:
Sold.